### PR TITLE
Support regex in feign client @RequestMapping

### PR DESCRIPTION
**Pull request by venilnoronha**
I've added `@RequestMapping` regex parsing to `SpringMvcContract` to now accept `@FeignClient` mappings such as `@RequestMapping("/{login:[_'.@a-z0-9-]+}")`. See #816 for more info.

**Sample Usage**
```
@FeignClient("service") // Can be a regular URL
public interface ServiceClient {

    @RequestMapping("/api/{symbolicName:[a-z-]+}-{version:\\d\\.\\d\\.\\d}{extension:\\.[a-z]+}")
    ResponseEntity<Model> fetch(@PathVariable("symbolicName") String symbolicName,
        @PathVariable("version") String version, @PathVariable("extension") String extension);

}
```

Please review and pull. I've signed the CLA.

Thanks,
Venil Noronha

### CODE REVIEW

**Comment by venilnoronha** [comment]
Build is failing now due to some heap space error.

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/feign/support/SpringMvcContract.java [hunk 1]
```diff
@@ -144,6 +154,25 @@ public MethodMetadata parseAndValidateMetadata(Class<?> targetType, Method metho
 		return md;
 	}
 
+	protected void parseRegexAndUpdatePath(MethodMetadata md, Method method) {
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/feign/support/SpringMvcContract.java]: Can you walk me through what this method is doing and why?

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/feign/support/SpringMvcContract.java [hunk 2]
```diff
@@ -144,6 +154,25 @@ public MethodMetadata parseAndValidateMetadata(Class<?> targetType, Method metho
...
+			for (int i = 0; i < paramCount; i++) {
+				MethodParameter param = new SynthesizingMethodParameter(method, i);
+				param.initParameterNameDiscovery(parameterNameDiscoverer);
+				if (param.hasParameterAnnotation(PathVariable.class)) {
+					PathVariable pathVariable = param.getParameterAnnotation(PathVariable.class);
+					String varName = pathVariable.value();
+					url = url.replaceAll("\\{" + varName + ":.+?\\}", "{" + varName + "}");
+				}
+			}
+			Field urlField = ReflectionUtils.findField(RequestTemplate.class, "url");
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/feign/support/SpringMvcContract.java]: This worries me a bit. Without having looked myself, is there a way to  do this without reflection?

File: spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/feign/support/SpringMvcContract.java [hunk 3]
```diff
@@ -144,6 +154,25 @@ public MethodMetadata parseAndValidateMetadata(Class<?> targetType, Method metho
...
 
+	protected void parseRegexAndUpdatePath(MethodMetadata md, Method method) {
+		if (isAnnotated(method, RequestMapping.class)) {
+			String url = md.template().url();
+			int paramCount = method.getParameterTypes().length;
+			for (int i = 0; i < paramCount; i++) {
+				MethodParameter param = new SynthesizingMethodParameter(method, i);
+				param.initParameterNameDiscovery(parameterNameDiscoverer);
+				if (param.hasParameterAnnotation(PathVariable.class)) {
+					PathVariable pathVariable = param.getParameterAnnotation(PathVariable.class);
```
Comment (spencergibb) [inline spring-cloud-netflix-core/src/main/java/org/springframework/cloud/netflix/feign/support/SpringMvcContract.java]: We went through steps to remove reference to `PathVariable` from this class. See `PathVariableParameterProcessor`.

**Comment by spencergibb** [comment]
Tell me if I'm wrong, but this doesn't 'support' regex's in path as ignores them. Am I correct?

**Comment by venilnoronha** [comment]
Yes. I thought the use case was to use a common `interface` for the `@Controller` and the `@FeignClient`, which one wasn't able to do `@FeignClient` with regex in path would throw a `IllegalArgumentException`. That's the reason I thought of ignoring the regex part of the path.

If this is not what we're looking at I can rework this.

**Comment by dsyer** [comment]
One of the goals of Spring Cloud Netflix is *not* sharing an interface between a `@Controller` and a `@FeignClient`. So unless we can think of a way to actually support regex on the feign side, I don't see much advantage in adding this feature.

**Comment by spencergibb** [comment]
Closing per our comments. @venilnoronha please feel free to open a new PR if you want to support regex's, not ignore them.

**Pull request closed without merging**